### ISSUE TITLE
Settings comparison with WordPress.com

### ISSUE THREAD

**Issue by joehoyle** [WP-API/WP-API#2860]
## Comparison of Settings in WordPress.com

The response format for the WordPress.com API is somewhat different and follows the pattern:

``` json
{
    "ID": 123,
    "name": "My Site"
    ...,
    "settings": {
        "admin_url" : "https://....",
        "require_name_email": false,
        ...
    }
}
```

The WP REST API Response has:

``` json
{
    "url" : "https://...",
    "default_category": 1,
    ...
}
```

It's not clear _why_ the WordPress.com API takes this structure with a `settings` sub key, however it would be good to find out why!

Broadly, I'd say the settings names and values in the WordPress.com API have had little change, and therefor can be somewhat cryptic with names such as `wga`, `holidaysnow`, `comment_whitelist` - as there's no published documentation for the Settings API, it seems rather difficult for a developer to use (however I think this is being worked on).

Updating settings is also different on the WordPress.com api, instead of it returning the updated resource, it returns an object with which settings for were updated:

``` json
{ "updated": [] }
```

To update settings, you don't send the same object shape received in the GET request, you instead send a key => value object. This matches the style in the WP REST API, however specific option names may differ.

``` json
{ "default_ping_status" : "open" }
```
### What we can adopt

I think the best thing we can learn from the WordPress.com Settings API is the fields they have chosen in include in the response. There's a mixture of core and jetpack settings, so we should of course ignore the jetpack ones, but make all the core ones available via WP REST API.

This will provide like-for-like _functionality_ (though not an exact map) for WordPress core features
to the WordPress.com API.

I recommend we _don't_ follow the same response format with the `settings` sub key, as there doesn't appear to be a good enough reason for this abnormal structure. Likewise, I recommend we _don't_ follow the same response format when updating settings with the `updated` key, as we should stay consistent with all other REST endpoints which is to return the same formatted object as a `GET` request.

**Comment by tobeycodes** [comment]
The settings sub key doesn't make much sense to me either. ID and Name are still considered settings to me, usually they won't change but they can.

**Comment by nylen** [comment]
cc @rralian who wrote the initial version of the WP.com settings endpoint for questions on response format.

**Comment by nylen** [comment]
It makes sense for us to add our Jetpack-specific settings via `register_setting`.

Looking at the settings registered [in `plugin.php`](https://github.com/WP-API/WP-API/blob/c57ec7a/plugin.php#L199), there are a couple that we'll need to either remove or make read-only (`url` and perhaps `email`).

It's also not hard to envision use cases for this endpoint where the option value is not stored as a site option.  I propose adding a couple of filters in `WP_REST_Settings_Controller::get_item` and `::update_item` that allow you to override getting and updating settings respectively.  I'll work up a PR.

**Comment by rmccue** [comment]
cc @m who asked specifically about this. :)

**Comment by rralian** [comment]
At the time we didn't have any method in the REST API that would allow changing Site settings. All we had was the Site endpoint, which returned a smaller set of data. And I didn't want to change the Site endpoint to return every setting we supported, because we had to keep the Site object small so the `/me/sites` endpoint could be as fast as possible. Besides, most of these settings aren't too relevant to Calypso unless you want to change the setting, so they aren't necessary unless you're on the settings page anyway. So that led to the /sites/$site/settings endpoint itself.

I needed to support the ability to GET/POST some of those root Site attributes as well as a whole slew of new settings. For example, the "general settings" page needed to support changing a mix of these (both Site Title and Related Posts).  It made sense to me at the time to keep these new settings somewhat separate from the root attributes you would otherwise get from the Site endpoint. This also mapped somewhat to our more object-oriented approach to data at the time, where those root attributes would map to the existing Site object, and the data under the `settings` key was stored/updated in a separate `settings` area of the Site object in Calypso. You'd have a `Site` object in Calypso and you'd call `Site.getSettings()` to retrieve these extra settings.

Short version, there's no need to maintain that.

As you're looking at this, one thing you might want to think about is how big this settings endpoint is going to get. It wasn't clear to me at the time how big the settings endpoint was going to get and I considered subdividing settings into specific pages. Like "writing" settings and "reading" settings. But ultimately went with a single fetch because that seemed fast enough. But the WP.com settings endpoint is still missing quite a few of the settings that we otherwise support in wp-admin for both WP.com and .org sites.

**Comment by aduth** [comment]
> can be somewhat cryptic with names such as `wga`, `holidaysnow`, `comment_whitelist` - as there's no published documentation for the Settings API, it seems rather difficult for a developer to use

As an aside, while not as friendly as the developer site, field information for all endpoints (including undocumented) can be found in Jetpack's [`json-endpoints.php` file](https://github.com/Automattic/jetpack/blob/master/json-endpoints.php)*. The settings fields, for example:

https://github.com/Automattic/jetpack/blob/1801d53/json-endpoints.php#L2443-L2495

_(Link pinned to current version to preserve line numbers)_

_*_ Some Jetpack-specific endpoints in [`json-api-jetpack-endpoints.php`](https://github.com/Automattic/jetpack/blob/master/json-endpoints/jetpack/json-api-jetpack-endpoints.php)

**Comment by rmccue** [comment]
@joehoyle Can you own moving this to Trac?

**Comment by joehoyle** [comment]
Closing in favor of https://core.trac.wordpress.org/ticket/38490 also added a patch there.
